Test search page, creating tests

grunwjir opened the pull request:
Test search page was improved, it supports list and card view. UI changes. 

The page for test detail was created, it is possible to add/edit/remove metrics and alerts. Subscribing to test alerts is working. UI of this page is temporary, will be changed.

Form validation for tests, metrics and alerts was implemented.

grunwjir committed:
[frontend - angular] Updated test detail, it's possible to create/edit/remove metrics and alerts. UI for test detail is only illustrative, will be redesigned.

grunwjir committed:
[dummy service layer] Added validation in tests, test alerts and test metrics. [enum OrderBy] Renamed order by option for group property (from GROUP_ID_ASC to GROUP_ASC)

grunwjir committed:
[angular] Created component for pagination - it allows to change page size and set current page. [form validation] Added validation for test, alert and metric.

grunwjir committed:
[frontend] Refactoring. Changed directory structure for test, metric moved to app folder.

grunwjir committed:
[frontend] Test search was improved. Created list and card view component for the test.

grunwjir committed:
[frontend] Progress bar to page header was added. Search page for tests was improved - pagination style, load tests before the page is rendered. Added subscribing to test alerts.

grunwjir committed:
[frontend] Refactoring - Added metric and alert service. It contains methods for creating forms in a modal window.

grunwjir commented on model/src/main/java/org/perfrepo/model/to/OrderBy.java:
```diff
@@ -27,8 +27,8 @@
 
-   GROUP_ID_DESC;
+   GROUP_DESC;
 
```
Can I rename this?

Holmistr replied:
It's actually better if you rename this :)

grunwjir committed:
[frontend build] Added frontend building support - gulp. Node, npm, bower is locally downloaded by maven. Frontend library dependencies are automatically resolved. All script, template and style files was moved to webapp_content folder. Gulp task processes these files and copy them to right webapp folder.

grunwjir committed:
[dummy service layer] Created DTO builders. It's easier to create instances of DTO objects. Use only for dummy service implementation.

grunwjir committed:
[test execution] Completed TestExecutionDto, DTO object builder for TestExecutionDto was created. Created rest endpoint and adapter for test execution.

grunwjir committed:
[test execution - front] Basic structure (controller, service, view) for test execution was created.

grunwjir committed:
[test execution - in progress] Added test execution search page. Creating and editing test executions.

grunwjir committed:
[checkstyle plugin] Resolved errors. [gulp] Added dependency.

grunwjir committed:
[removed webapp files] These files were moved to "webapp_content" directory and are copied to "webapp" by gulp build.

Holmistr commented on model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionGroupValueDto.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * @author Jiri Grunwald (grunwjir@gmail.com)
+ */
+public class TestExecutionGroupValueDto {
```
Could you please explain purpose of this class? I would also rename it to ValueGroupDto, you don't have to put "TestExecution" everywhere

grunwjir replied:
It represents a group of test execution values. This group is for each test metric, therefore the values are grouped by a metric.

Holmistr commented on model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionSearchParams.java:
```diff
@@ -1,10 +1,73 @@
  *
  * @author Jiri Grunwald (grunwjir@gmail.com)
  */
 public class TestExecutionSearchParams {
```
I think TestExecutionSearchCriteria is more descriptible, but that's a minor comment.

grunwjir replied:
Renamed

Holmistr commented on model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionValueDto.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * @author Jiri Grunwald (grunwjir@gmail.com)
+ */
+public class TestExecutionValueDto {
```
Rename to ValueDto. Values are present only for test execution, it's obvious.

grunwjir replied:
Ok, renamed.

Holmistr commented on web/pom.xml:
```diff
@@ -306,26 +306,48 @@
-            <artifactId>exec-maven-plugin</artifactId>
-            <version>1.3.2</version>
+            <groupId>com.github.eirslett</groupId>
+            <artifactId>frontend-maven-plugin</artifactId>
```
Could you please very shortly explain, what the plugin basically does? It's just I have never used it before, nothing wrong with the usage at all :)

grunwjir replied:
It downloads and installs Node, NPM locally for this project. It runs Bower and executes Gulp build task.

Holmistr commented on web/src/main/java/org/perfrepo/web/adapter/dummy_impl/builders/TestExecutionValueDtoBuilder.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * @author Jiri Grunwald (grunwjir@gmail.com)
+ */
+public class TestExecutionValueDtoBuilder {
```
Rename, see previous comments.

grunwjir replied:
Renamed.

Holmistr commented on web/src/main/java/org/perfrepo/web/rest/endpoints/TestExecutionRestApi.java:
```diff
@@ -0,0 +1,82 @@
+              .entity(result.getData()).build();
+   }
+
+
```
Extra empty line

Holmistr commented:
@grunwjir I've placed some minor comments, but let me add some more high-level regarding tests pages:
* the listing is not expanding to full width on my monitor, looks like the width is fixed somehow and it looks IMHO weird (see attached screenshot)
* add headings to the columns of the table (like test name, test UID)
* the test name should be clickable on its own redirecting to the test detail
* There is button "Test detail". First - it can be just "Detail" with maybe different icon, but I would also create buttons "Edit" and "Delete" (also remove "test", it's obvious) and now hide them into dropdown menu. However, I really like the dropdown menu and the "Create test execution" and "Show test executions" should remain there, that's absolutely brilliant :)
* there is a small arrow on each row showing more information. To me it doesn't make sense to duplicate the information about the test there. However, I really like the idea of this dropdown and I think we can use it much better. It could show last x (let's say 3) test executions that were run for this test. Just a mini-table with 3 rows saying something like "Last 3 test executions" and listing - execution name, execution date and tags. That would be truly awesome, hopefully it's not too much work. And it's important that this information would be lazily-loaded, not preloaded for every test in advance. 
* Also the arrow is not vertically centered with the test name, so it looks a bit weird.
* regarding the card view - I cannot imagine a usecase how it could be useful in this situation, hence I would remove it

Moving on to test detail page:
* the right align of "Uid", "Group" etc. is done on purpose? Looks non-traditional to me :-D I would prefer left align. Also please use UID everywhere :)
* listings of metric should contain also the description of the metric, it's actually quite common that you look at that to find out, what it actually means.
* listings of alerts - basically the same as above, I would show name of the alert and description, everything else can be hidden until "Edit" is clicked.
* even here (test detail) the page is narrower on my screen, perhaps something more general is not set up correctly (see attached screenshot)

That's all I've got. I know that some of the comments are really minor, please don't take it bad, I think it's an awesome work, just want to have it perfect :)

![snimek z 2017-01-11 13-35-29](https://cloud.githubusercontent.com/assets/4253850/21848806/0921ee8c-d803-11e6-9a43-f0112b164f6e.png)
![snimek z 2017-01-11 13-35-33](https://cloud.githubusercontent.com/assets/4253850/21848809/0bcb7e64-d803-11e6-9af1-16da4594f4db.png)

grunwjir committed:
[refactoring] Rename some classes.

Prefix "TestExecution" was removed from name classes where it were unnecessary. Other small changes (removed extra empty lines...)

grunwjir committed:
[frontend] Set the page content width to full width of the screen.

The max width 1140px of the page content was removed.

grunwjir committed:
[frontend] Remove the card view in tests and test executions search page.

The card view was removed. It was useless at that moment.

grunwjir committed:
[frontend] Remove possibility to show a row detail in the list view

It contained duplicate information about tests and test executions. The good idea is to show last x test executions for a test row (oncoming feature).

grunwjir committed:
[frontend] Add and adjust action buttons in the row of list view.

Action buttons in the row of list view for tests search and test executions search was added and adjusted.

grunwjir committed:
[frontend] Remove colons in form labels. Rename Uid to UID.

grunwjir committed:
[frontend] Add link to test name in list view.

It redirects to the test detail. The same behavior is for test execution.

grunwjir committed:
[backend] Add javadoc comments.

grunwjir commented:
Answers to your comments:
> - the listing is not expanding to full width on my monitor, looks like the width is fixed somehow and 
it looks IMHO weird (see attached screenshot)

Repaired to full width of a screen. It was set to maximum width 1140px, but it was bad idea. 

> - add headings to the columns of the table (like test name, test UID)

It is [list view component](https://www.patternfly.org/pattern-library/content-views/list-view/#/_design) and it does not support headings. I know that my solution is not ideal, will be improved.

> - the test name should be clickable on its own redirecting to the test detail
> - There is button "Test detail". First - it can be just "Detail" with maybe different icon, but I would also create buttons "Edit" and "Delete" (also remove "test", it's obvious) and now hide them into dropdown menu. However, I really like the dropdown menu and the "Create test execution" and "Show test executions" should remain there, that's absolutely brilliant :)

Ok, resolved.

> - there is a small arrow on each row showing more information. To me it doesn't make sense to duplicate the information about the test there. However, I really like the idea of this dropdown and I think we can use it much better. It could show last x (let's say 3) test executions that were run for this test. Just a mini-table with 3 rows saying something like "Last 3 test executions" and listing - execution name, execution date and tags. That would be truly awesome, hopefully it's not too much work. And it's important that this information would be lazily-loaded, not preloaded for every test in advance.
>  - Also the arrow is not vertically centered with the test name, so it looks a bit weird.

I agree and show last x test executions is great idea, I will add it (next PR).

> - regarding the card view - I cannot imagine a usecase how it could be useful in this situation, hence I would remove it

The card view was removed.

> Moving on to test detail page:
...

Test detail page will be completely redesigned, next PR.

Thank you very much for your revision :)

Holmistr commented:
Changes looks good to me, I have noticed two more things:
* I would also want to see to which group the tests is assigned in the listings, hence add another column like "Group".
* I noticed that when I deploy the application to the server and access the "base" URL, the dashboard can be seen without login. However when I want to access anything else, login is finally required. This is something that we must not forget to fix :)

Please add at least the group column, note somewhere the login thing and we're good to integrate this PR.

grunwjir committed:
[frontend] Add redirect to login page for unauthenticated user

When the app is loaded for the first time and a user is unauthenticated, then the user is redirected to login page.

grunwjir committed:
[frontend] Add group name column to the test listing view
